libclamav_rust/src/fuzzy_hash.rs

```diff
@@ -0,0 +1,544 @@
+// Sample c-facing function call
+
+use image::{imageops::FilterType::Lanczos3, DynamicImage, ImageBuffer, Luma, Pixel, Rgb};
+use num_traits::{NumCast, ToPrimitive, Zero};
+use rustdct::DctPlanner;
+use transpose::transpose;
+
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::ffi::CStr;
+use std::mem::ManuallyDrop;
+use std::os::raw::c_char;
+use std::slice;
+
+use log::*;
+use thiserror::Error;
+
+use crate::sys;
+use crate::validate_str_param;
+
+/// CdiffError enumerates all possible errors returned by this library.
+#[derive(Error, Debug)]
+pub enum FuzzyHashError {
+    #[error("Invalid format")]
+    Format,
+
+    #[error("Unknown algorithm: {0}")]
+    UnknownAlgorithm(String),
+
+    #[error("Invalid hamming distance: {0}")]
+    FormatHammingDistance(String),
+
+    #[error("Invalid hash: {0}")]
+    FormatHashBytes(String),
+
+    #[error("Failed to load image: {0}")]
+    ImageLoad(image::ImageError),
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+pub enum FuzzyHash {
+    Image([u8; 8]),
+}
+
+impl std::fmt::Display for FuzzyHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            &FuzzyHash::Image(hash_bytes) => {
+                write!(f, "{}", hex::encode(hash_bytes))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct FuzzyHashMeta {
+    lsigid: u32,
+    subsigid: u32,
+    _hamming_distance: u32,
+}
+
+/// Initialize the hashmap
+#[no_mangle]
+pub extern "C" fn fuzzy_hash_init_hashmap() -> sys::hashmap_ptr_t {
```

Comment:
> A general comment on the design:
> 
> You're giving Rust HashMap objects directly back to the caller. However, you _could_ be giving back a `struct` that wraps the HashMap. Then, functions like `fuzzy_hash_load_subsignature` would drop into an `impl` for that structure and take `&self`. I think it would end up being a lot cleaner, and gives you a little more flexibility later.

I think I've made all of the suggested changes except this one. We could talk more about what you mean. I'm otherwise ready for re-review.